src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java

```diff
@@ -0,0 +1,375 @@
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * This class is thread-safe
+ *
+ * @author <a href="http://tfox.org">Tim Fox</a>   7                                                                                     T
+ */
+public class ClusteredEventBus extends LocalEventBus {
+
+  private static final Logger log = LoggerFactory.getLogger(ClusteredEventBus.class);
+
+  private static final Buffer PONG = Buffer.buffer(new byte[] { (byte)1 });
+  private static final String SERVER_ID_HA_KEY = "server_id";
+  private static final String SUBS_MAP_NAME = "__vertx.subs";
+
+  private final ClusterManager clusterManager;
+  private final HAManager haManager;
+  private final ConcurrentMap<ServerID, ConnectionHolder> connections = new ConcurrentHashMap<>();
+  private final Context sendNoContext;
+  private VertxOptions options;
+  private AsyncMultiMap<String, ServerID> subs;
+  private ServerID serverID;
+  private NetServer server;
+
+  public ClusteredEventBus(VertxInternal vertx,
+                           VertxOptions options,
+                           ClusterManager clusterManager,
+                           HAManager haManager) {
+    super(vertx);
+    this.options = options;
+    this.clusterManager = clusterManager;
+    this.haManager = haManager;
+    this.sendNoContext = vertx.getOrCreateContext();
+    setNodeCrashedHandler(haManager);
+  }
+
+  @Override
+  public void start(Handler<AsyncResult<Void>> resultHandler) {
+    clusterManager.<String, ServerID>getAsyncMultiMap(SUBS_MAP_NAME, ar2 -> {
+      if (ar2.succeeded()) {
+        subs = ar2.result();
+        server = vertx.createNetServer(new NetServerOptions().setPort(options.getClusterPort()).setHost(options.getClusterHost()));
+        server.connectHandler(getServerHandler());
+        server.listen(asyncResult -> {
+          if (asyncResult.succeeded()) {
+            int serverPort = getClusterPublicPort(options, server.actualPort());
+            String serverHost = getClusterPublicHost(options);
+            serverID = new ServerID(serverPort, serverHost);
+            haManager.addDataToAHAInfo(SERVER_ID_HA_KEY, new JsonObject().put("host", serverID.host).put("port", serverID.port));
+            if (resultHandler != null) {
+              started = true;
+              resultHandler.handle(Future.succeededFuture());
+            }
+          } else {
+            if (resultHandler != null) {
+              resultHandler.handle(Future.failedFuture(asyncResult.cause()));
+            } else {
+              log.error(asyncResult.cause());
+            }
+          }
+        });
+      } else {
+        if (resultHandler != null) {
+          resultHandler.handle(Future.failedFuture(ar2.cause()));
+        } else {
+          log.error(ar2.cause());
+        }
+      }
+    });
+  }
+
+  @Override
+  public void close(Handler<AsyncResult<Void>> completionHandler) {
```

Comment:
EventBus#close javadoc says _completionHandler may be {@code null}_ : either change javadoc or check completionHandler is not null - same for parent class.